calculus.py

```diff
@@ -292,3 +294,88 @@ def secant_wrapper(func, x0, x1, args=(), maxiter=50):
         "converged": res.converged,
         "iterations": res.iterations,
         "function_calls": res.function_calls}
+
+# Root Finding with Bisection Method
+def bisection_wrapper(func, a, b, tol=1e-6, max_iter=1000):
+    """
+    Wrapper for SciPy's `bisect` function.
+
+    Parameters:
+        func: The function for which to find the root.
+        a: The start of the interval.
+        b: The end of the interval.
+        tol: The tolerance level for convergence. Defaults to 1e-6.
+        max_iter: Maximum number of iterations. Defaults to 1000.
+
+    Returns:
+        Root: The approximate root of the function.
+
+    Raises:
+        ValueError: If func(a) and func(b) do not have opposite signs or if
+                    the function encounters undefined values (singularities).
+    """
+    small_value_threshold = 1e-3  # Threshold for detecting singularities in sin(x)
+
+    try:
+        # Check if sin(a) or sin(b) are very small (near zero)
+        if abs(math.sin(a)) < small_value_threshold:
+            raise ValueError(f"Singularity detected: division by zero in function at x = {a}.")
+        if abs(math.sin(b)) < small_value_threshold:
+            raise ValueError(f"Singularity detected: division by zero in function at x = {b}.")
+
+        # Call the SciPy bisect method if no errors were raised
+        root = bisect(func, a, b, xtol=tol, maxiter=max_iter)
+
+    except ValueError as e:
+        raise ValueError(f"SciPy bisect failed: {e}") from e
+
+    return root
+
+def bisection_pure_python(func, a, b, tol=1e-6):
```

Comment:
as in scipy case a timeout condition would be useful

Reply:
These changes are in first PR, which is already merged.